### PR TITLE
fix: don't focus on composer when open right modal

### PR DESCRIPTION
**Pull request by dummy-1111**
### Fixed Issues

$ https://github.com/Expensify/App/issues/26489
PROPOSAL: https://github.com/Expensify/App/issues/26489#issuecomment-1703141465

### Tests
1. Open ND and log in with any account
2. Open any chat
3. Input `@` character and verify that mention suggestions appears
4. Tap `+` button again and tap any menu except `Add attachment`
5. Verify that suggestion list doesn't appear for a brief moment before modal is opened

- [x] Verify that no errors appear in the JS console

### Offline tests
1. Open ND and log in with any account
2. Open any chat
3. Input `@` character and verify that mention suggestions appears
4. Tap `+` button again and tap any menu except `Add attachment`
5. Verify that suggestion list doesn't appear for a brief moment before modal is opened

### QA Steps
1. Open ND and log in with any account
2. Open any chat
3. Input `@` character and verify that mention suggestions appears
4. Tap `+` button again and tap any menu except `Add attachment`
5. Verify that suggestion list doesn't appear for a brief moment before modal is opened

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/126839717/84e3b74f-7fff-4b24-a073-89359373ef1d

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/126839717/89363ba5-2f15-4c6c-ba93-479eb2a0b654

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/126839717/c2b9e580-df24-4b08-bfda-be9baffc5ecb

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/126839717/217c648e-cc18-47ee-b1a1-18695ab5cf36

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/126839717/0baf1a0e-a1cf-4eda-932e-0f0c5561b029

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/126839717/d80c8357-d39b-41cd-8489-ba0b8fec1215

</details>

### CODE REVIEW

File: src/pages/home/report/ReportActionCompose/ReportActionCompose.js [hunk 1]
```diff
@@ -179,11 +181,17 @@ function ReportActionCompose({
             return;
         }
         composerRef.current.focus(true);
+
+        // Restore the suggestions if onSelectionChange is not called on focused
+        if (isSelectionChangedOnFocus || !suggestionsRef.current) {
+            return;
+        }
+        suggestionsRef.current.restoreSuggestions();
```
Comment (s77rt) [inline src/pages/home/report/ReportActionCompose/ReportActionCompose.js:189]: What is this for?

**Comment by s77rt** [comment]
@s-alves10 The PR is slightly different than the approved solution. Please comment on each new section that was added explaining the need for it.

File: src/pages/home/report/ReportActionCompose/ReportActionCompose.js (hunk 1 above)
Comment (s-alves10) [reply src/pages/home/report/ReportActionCompose/ReportActionCompose.js:189]: `onSelectionChange` isn't called when composer gets focused(without any selection changes) on native platforms
In this case, we can restore the previous suggestion list. Previous suggestions are stored in `resetSuggestions` on blur handler

File: src/pages/home/report/ReportActionCompose/ReportActionCompose.js [hunk 2]
```diff
@@ -306,6 +319,13 @@ function ReportActionCompose({
         [],
     );
 
+    useEffect(() => {
+        if (modal.isVisible) {
+            return;
+        }
+        focus();
+    }, [modal.isVisible]);
```
[the remark this answers is not in the document]
Comment (s-alves10) [reply src/pages/home/report/ReportActionCompose/ReportActionCompose.js:327]: Composer doesn't get focused when right modal is closed on native platforms.
This is reproducible on staging
This was added in order to refocus on modal hide

File: src/pages/home/report/ReportActionCompose/ReportActionCompose.js [hunk 3]
```diff
@@ -179,11 +181,17 @@ function ReportActionCompose({
...
+        if (isSelectionChangedOnFocus || !suggestionsRef.current) {
+            return;
+        }
+        suggestionsRef.current.restoreSuggestions();
     };
 
     const isKeyboardVisibleWhenShowingModalRef = useRef(false);
     const restoreKeyboardState = useCallback(() => {
-        if (!isKeyboardVisibleWhenShowingModalRef.current) {
+        if (!isKeyboardVisibleWhenShowingModalRef.current || !shouldFocusInputOnScreenFocus) {
```
Comment (s-alves10) [inline src/pages/home/report/ReportActionCompose/ReportActionCompose.js:194]: We don't need to set focus manually if focus is set automatically

**Comment by s-alves10** [comment]
> @s-alves10 The PR is slightly different than the approved solution. Please comment on each new section that was added explaining the need for it.

I added some comments.
`Suggestions`, `SuggestionEmojis`, and `SuggestionMentions` have a new method: `restoreSuggestions`. This is created for restoring the previous suggestions.
We calculate the suggestions in `onSelectionChange`, but this is not called when gets refocused if there is no selection change. In this case, I restored the previous suggestions.

**Comment by s-alves10** [comment]
@s77rt 

Please take a look again

File: src/pages/home/report/ReportActionCompose/ReportActionCompose.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (s77rt) [reply src/pages/home/report/ReportActionCompose/ReportActionCompose.js:327]: This is probably intended. The keyboard that shows on focus on native is annoying so auto focus is not something wanted here. Can you please revert this and other changes related to this?

[the remark this answers is not in the document]
Comment (s-alves10) [reply src/pages/home/report/ReportActionCompose/ReportActionCompose.js]: @s77rt 

Reverted

File: src/pages/home/report/ReportActionCompose/ReportActionCompose.js (hunk 1 above)
Comment (s77rt) [reply src/pages/home/report/ReportActionCompose/ReportActionCompose.js:187]: Is this still nedeed?

File: src/pages/home/report/ReportActionCompose/ReportActionCompose.js (hunk 3 above)
Comment (s77rt) [reply src/pages/home/report/ReportActionCompose/ReportActionCompose.js:192]: What problem is this solving?

**Comment by s77rt** [comment]
Added some comments on the previous review comments

File: src/pages/home/report/ReportActionCompose/ReportActionCompose.js (hunk 1 above)
Comment (s-alves10) [reply src/pages/home/report/ReportActionCompose/ReportActionCompose.js:187]: Without this, suggestions won't be shown unless selection is changed in native platforms. Refocusing or mouse clicking doesn't work.

File: src/pages/home/report/ReportActionCompose/ReportActionCompose.js (hunk 3 above)
Comment (s-alves10) [reply src/pages/home/report/ReportActionCompose/ReportActionCompose.js:192]: We have composer focused already if `shouldFocusInputOnScreenFocus` is true. In android chrome, onSelectionChange isn't called when we set focus again.
You can reproduce this with the following steps(Android chrome):
1. Open any chat and input `@` to show mentions list
2. Click + button to open menu and tap outside of the menu to close it
3. Verify that mention list doesn't show

**Comment by s-alves10** [comment]
@s77rt 

If you think we should limit the scope of the PR in the issue, please let me know.
We don't need `restoreSuggestions` in that case. I'll remove them

**Comment by s77rt** [comment]
@s-alves10 Yes, let's please limit the PR to the bugs that are linked in the issue. If a bug is found and it's a regression we should fix it, but if it's on staging/main already it's better to report it in Slack otherwise we would be working out a new solution instead of going forward with the approved one (with the exception of some bugs where the solution is straightforward).

File: src/pages/home/report/ReportActionCompose/AttachmentPickerWithMenuItems.js [hunk 4]
```diff
@@ -257,6 +261,7 @@ function AttachmentPickerWithMenuItems({
                             onClose={onPopoverMenuClose}
                             onItemSelected={(item, index) => {
                                 setMenuVisibility(false);
+                                onItemSelected();
```
Comment (s77rt) [inline src/pages/home/report/ReportActionCompose/AttachmentPickerWithMenuItems.js:264]: Can we move this to be the last function call?

**Comment by s-alves10** [comment]
@s77rt 

I removed all unrelated code. Please check again

**Review by s77rt — CHANGES_REQUESTED** [review]

File: src/pages/home/report/ReportActionCompose/AttachmentPickerWithMenuItems.js (hunk 4 above)
Comment (s-alves10) [reply src/pages/home/report/ReportActionCompose/AttachmentPickerWithMenuItems.js:264]: We need to overwrite `isKeyboardVisibleWhenShowingModalRef.current` value in `triggerAttachmentPicker` function when selecting `Add attachment`.
I think this is the ideal position this call should be placed.
Is there any specific reason for moving?

**Review by s77rt — APPROVED** [review]

**Comment by s-alves10** [comment]
@youssef-lr 

Please take a look

**Review by youssef-lr — APPROVED** [review]
LGTM!

**Pull request merged**